### PR TITLE
Perturbed oracles with variance reduction

### PR DESCRIPTION
**Pull request by BatyLeo**
This PR reworks all the `AbstractPerturbed` related structs:
- Most methods (including the rrule) are now generic and implemented directly on `AbstractPerturbed`
- As a kwarg option enabled by default, gradients can be computed with variance reduction
- Additive and multiplicative perturbations still exist as a special case, and are compatible with Fenchel-Young losses
- Custom distributions can be used for additive and multiplicative perturbations
- New `PerturbedOracle` layer, generalizing additive/multiplicative perturbations, where the perturbation can have any form. It cannot be used with Fenchel-young losses
- Tests for new options
- New tests checking that a `PerturbedOracle` with the correct distribution coincides with `PerturbedAdditive` and `PerturbedMultiplicative`

### CODE REVIEW

**Comment by gdalle** [comment]
I won't have time to review this, and I probably need to take a step back to focus on my own research. It would be nice if you and @LouisBouvier could rely on each other for PR reviewing, can you bring him up to speed?

**Comment by gdalle** [comment]
As far as custom distributions are concerned, it would be nice but we need a gradlogpdf function if memory serves me well. This is not necessarily easy to find

**Comment by gdalle** [comment]
And I agree that Fenchel Young is messy. This implementation tried to minimize code duplication, but possibly at the expense of clarity.

**Comment by BatyLeo** [comment]
> As far as custom distributions are concerned, it would be nice but we need a gradlogpdf function if memory serves me well. This is not necessarily easy to find

The gradlogpdf can either be given by the user or computed with autodiff, this is already the case in the new generic `PerturbedOracle` struct, but it cannot be used with `FenchelYoungLoss` at the moment

**Comment by gdalle** [comment]
> The gradlogpdf can either be given by the user or computed with autodiff

Will we have a problem with nested autodiff?

**Comment by gdalle** [comment]
Maybe to avoid random failures across Julia versions we could use StableRNGs in testing?

**Comment by BatyLeo** [comment]
> > The gradlogpdf can either be given by the user or computed with autodiff
> 
> Will we have a problem with nested autodiff?

What do you mean by "nested autodiff" ?



> Maybe to avoid random failures across Julia versions we could use StableRNGs in testing?

I'm still investigating, but it seems that even whithin the same Julia versions our tests are not deterministic. The failing test passes on my laptop about half the time

**Comment by gdalle** [comment]
> What do you mean by "nested autodiff" ?

Differentiating something with Zygote that has already been differentiated with it once. I don't think it will be the case here

File: src/perturbed/abstract_perturbed.jl [hunk 1]
```diff
@@ -70,18 +69,64 @@ Turn random perturbations of `θ` into a distribution on polytope vertices.
...
+    M = get_nb_samples(perturbed)
+    function perturbed_oracle_dist_pullback(δy_dist)
+        weights = y_dist.weights
+        δy_weights = δy_dist.weights
+        δy_sum = sum(δy_weights)
+        δθ = sum(
+            map(1:M) do i
+                δyᵢ, ∇logpᵢ, w = δy_weights[i], ∇logp_samples[i], weights[i]
+                if autodiff_variance_reduction
+                    bᵢ = M == 1 ? 0 * δy_sum : (δy_sum - δyᵢ) / (M - 1)
```
Comment (LouisBouvier) [inline src/perturbed/abstract_perturbed.jl:121]: Do we want to average with respect to the w distribution or as we did in the former rrules for perturbed additive and multiplicative with equal weights on the atoms ?

**Comment by BatyLeo** [comment]
Should we enable variance reduction by default ? Currently it is not.

**Comment by gdalle** [comment]
We're gonna tag a breaking release anyway, we might as well

File: src/perturbed/multiplicative.jl [hunk 2]
```diff
@@ -1,80 +1,130 @@
 """
-    PerturbedMultiplicative <: AbstractPerturbed
+    PerturbedMultiplicative{P,G,O,R,S,parallel} <: AbstractPerturbed{parallel}
 
-Differentiable log-normal perturbation of a black-box maximizer: the input undergoes `θ -> θ ⊙ exp[εZ - ε²/2]` where `Z ∼ N(0, I)`.
+Differentiable log-normal perturbation of a black-box oracle:
```
Comment (LouisBouvier) [inline src/perturbed/multiplicative.jl:4]: Is it still log-normal since we have a generic distribution ?

**Comment by LouisBouvier** [comment]
Can we use variance reduction also in FY loss ? We also estimate an expectation by Monte-Carlo, when we use more than one sample we could think about it ?

**Comment by gdalle** [comment]
Since questions are popping up on the Julia forum, is this ready to go?

**Comment by BatyLeo** [comment]
> Since questions are popping up on the Julia forum, is this ready to go?

I think it is, Louis already reviewed it.
Maybe check the `AbstractPerturbed` rrule once more ?

**Comment by gdalle** [comment]
I trust you guys, you do the merge this time ;)

**Comment by gdalle** [comment]
Congrats, this was huge work!

**Pull request merged**